modules/dispatch/dispatch.go

```diff
@@ -0,0 +1,266 @@
+//
+// Contributor: Aaron Meihm ameihm@mozilla.com [:alm]
+
+// Package dispatch implements alert dispatching for the agent as a module.
+// Persistent modules which generate alerts will have these alerts forwarded
+// to this module if the dispatch module is active. The dispatch module can then
+// forward the alerts on based on it's configuration.
+package dispatch /* import "mig.ninja/mig/modules/dispatch" */
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"runtime"
+
+	"mig.ninja/mig"
+	"mig.ninja/mig/modules"
+)
+
+type module struct {
+}
+
+// NewRun returns a new instance of a modules.Runner for this module.
+func (m *module) NewRun() modules.Runner {
+	return new(run)
+}
+
+func init() {
+	modules.Register("dispatch", new(module))
+}
+
+type run struct {
+	Parameters Parameters
+	Results    modules.Result
+}
+
+func buildResults(e elements, r *modules.Result) (buf []byte, err error) {
+	r.Success = true
+	r.Elements = e
+	r.FoundAnything = true
+	buf, err = json.Marshal(r)
+	return
+}
+
+var logChan chan string
+var alertChan chan string
+var handlerErrChan chan error
+var configChan chan modules.ConfigParams
+
+// We keep a copy of the agent environment, hostname, and tags here so we can
+// use them in the dispatched output
+var env mig.AgentEnv
+var tags map[string]string
+var agtHostname string
+
+// messageBuf is a queue used to store incoming messages, and is drained by
+// runDispatch
+var messageBuf chan string
+
+// Dispatch record describes the formatting of JSON data submitted from the dispatch
+// module.
+type DispatchRecord struct {
+	Hostname string            `json:"hostname"`
+	Env      mig.AgentEnv      `json:"environment"`
+	Tags     map[string]string `json:"tags"`
+	Details  interface{}       `json:"details"`
+}
+
+func (d *DispatchRecord) fromString(msg string) error {
+	d.Hostname = agtHostname
+	d.Env = env
+	d.Tags = tags
+	err := json.Unmarshal([]byte(msg), &d.Details)
+	return err
+}
+
+func moduleMain() {
+	var cfg config
+
+	incfg := <-configChan
+	agtHostname = incfg.Hostname
+	tags = incfg.Tags
+	buf, err := json.Marshal(incfg.Env)
+	if err != nil {
+		handlerErrChan <- err
+		return
+	}
+	err = json.Unmarshal(buf, &env)
+	if err != nil {
+		handlerErrChan <- err
+		return
+	}
+	buf, err = json.Marshal(incfg.Config)
+	if err != nil {
+		handlerErrChan <- err
+		return
+	}
+	err = json.Unmarshal(buf, &cfg)
+	if err != nil {
+		handlerErrChan <- err
+		return
+	}
+	logChan <- "module received configuration"
+
+	messageBuf = make(chan string, 1024)
+
+	// Register the dispatch function, which will be called when the module
+	// recieves an alert message from the agent.
+	modules.RegisterDispatchFunction(dispatchIn)
+
+	// Start the dispatch dequeue function.
+	err = runDispatch(cfg)
+	if err != nil {
+		handlerErrChan <- err
+		return
+	}
+}
+
+func dispatchIn(msg string) {
+	select {
+	case messageBuf <- msg:
+	default:
+		// If we can't queue the message it is just dropped
```

Comment:
shouldnt that be logged?

Reply:
Agreed, I'll add some sort of feedback when this happens